### DISCUSSION TITLE
interrupted_signing

### DISCUSSION THREAD

**Discussion by nengxingxu** [MatthiasValvekens/pyHanko#154] [Q&A]
I use interrupted_signing.

int gen_digest, I calculate document summary, and save the object obj to the file file.

```
async def gen_digest(dic):
    prep_digest, signed_attrs, psi, output = await prep_doc(input_pdf, cert_path, timestamper_url, timestamper_timeout)
    with open('prep_digest.data', 'wb') as outf:
        pickle.dump(prep_digest, outf)
    with open('signed_attrs.data', 'wb') as outf:
        pickle.dump(signed_attrs, outf)    
    with open('psi.data', 'wb') as outf:
        pickle.dump(psi, outf)
    with open('output.data', 'wb') as outf:
        pickle.dump(output, outf)
```

in sign_pdf, read the object from the file, but it run crash. What am I going to do.

```
async def sign_pdf(dic):
    with open('prep_digest.data', 'rb') as inf:
        prep_digest = pickle.load(inf)
    with open('signed_attrs.data', 'rb') as inf:
        signed_attrs = pickle.load(inf)    
    with open('psi.data', 'rb') as inf:
        psi = pickle.load(inf)
    with open('output.data', 'rb') as inf:
        output = pickle.load(inf)
```

Traceback (most recent call last):
  File ".\interrupted_signing.py", line 263, in <module>
    main()
  File ".\interrupted_signing.py", line 257, in main
    outf.write(asyncio.run(sign_pdf(dic)).getvalue())
  File "C:\Python38\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Python38\lib\asyncio\base_events.py", line 616, in run_until_complete
    return future.result()
  File ".\interrupted_signing.py", line 230, in sign_pdf
    await proceed_with_signing(signing_certificate, prep_digest, signed_attrs, sig_value, new_psi, new_output)
  File ".\interrupted_signing.py", line 85, in proceed_with_signing
    await PdfTBSDocument.async_finish_signing(
  File "C:\Python38\lib\site-packages\pyhanko\sign\signers\pdf_signer.py", line 2237, in async_finish_signing
    await post_sign.post_signature_processing(
  File "C:\Python38\lib\site-packages\pyhanko\sign\signers\pdf_signer.py", line 2324, in post_signature_processing
    validation.DocumentSecurityStore.supply_dss_in_writer(
  File "C:\Python38\lib\site-packages\pyhanko\sign\validation\dss.py", line 445, in supply_dss_in_writer
    dss.register_vri(
  File "C:\Python38\lib\site-packages\pyhanko\sign\validation\dss.py", line 211, in register_vri
    ocsp_refs = set(
  File "C:\Python38\lib\site-packages\pyhanko\sign\validation\dss.py", line 133, in _cms_objects_to_streams
    obj_bytes = obj.dump()
  File "C:\Python38\lib\site-packages\asn1crypto\core.py", line 4124, in dump
    for index, (field_name, _, params) in enumerate(self._fields):
ValueError: not enough values to unpack (expected 3, got 2)

**Comment by MatthiasValvekens** [comment]
Hi @gudder,

Just from the pickle/unpickle invocations and the stack trace I can't tell you what's going on, but if you're doing everything right (and didn't accidentally switch around two function arguments or something like that), then it could be that there is some kind of incompatibility between `asn1crypto` and `pickle`. `asn1crypto` relies on some class initialisation voodoo, so I wouldn't be surprised if that would cause issues in a case like this.


If that's indeed what is going on, you have a couple of options:

 - When deserialising individual `asn1crypto` objects (like `signed_attrs` here), don't use pickle, but use `asn1crypto`'s `.dump()` and `.load()` functions instead. The whole purpose of `asn1crypto` is to provide a way to handle DER serialisation, so you might as well use the library for its intended purpose ;)
 - For things like `PostSigningInstructions`, you could manually implement a serialisation handler. Even if handling all the fields would be somewhat complicated, you probably only need a small subset for your use case to work.

Maybe, if `pickle` is extensible (and I don't know if it is), you could tell it to use `asn1crypto`'s own serialisation primitives to handle objects from `asn1crypto`.

------

I'm open to suggestions to improve the developer experience here, but I want to avoid depending on specific serialisation libraries (even standard ones like `pickle`). The intention of the interrupted signing workflow was to supply a way for an API user to _extract_ the necessary state to continue the signing process later on, but I left the serialisation process up in the air on purpose.

Anyway, if that's going to be a longer discussion, then it should probably get its own thread :).

**Reply by nengxingxu** [reply]
I have a question, that is, the python script is executed once to pickle dump and pickle load PostSigningInstructions, which is correct. If split twice, execute pickle once dump PostSigningInstructions. Pick the script for the second time Load PostSigningInstructions, which is incorrect.

**Reply by nengxingxu** [reply]
run the python script pickle dump for first time.
![image](https://user-images.githubusercontent.com/3445189/191162627-fdc7e978-5537-468d-9d66-29c21a951dfc.png)
run the python script second pickle load for second time.
![lQLPJxaz7e_h2vXNAenNAl2wsVlsxZwUiIoDJ9zaZMDOAA_605_489](https://user-images.githubusercontent.com/3445189/191163188-84babf3f-d980-4d69-ac73-d41f767cceb2.png)

**Reply by MatthiasValvekens** [reply]
Yep, that's what I thought. There's some bootstrapping logic in `asn1crypto` that rewrites those class properties, essentially to autofill the third tuple element. I guess that that doesn't run when you unpickle an asn1crypto object.

**Reply by MatthiasValvekens** [reply]
See my latest comment on the issue you raised on the `asn1crypto` repo for a workaround: https://github.com/wbond/asn1crypto/issues/239.

**Reply by MatthiasValvekens** [reply]
FYI, my PR to enable built-in `pickle` support in `asn1crypto` just got merged: wbond/asn1crypto#240. This should be fixed in the next release of `asn1crypto`.

**Reply by sbkis** [reply]
@nengxingxu did you get this to work with PostSignInstructions? It seems this has been solved, but I'm experiencing this exact issue for PostSignInstructions.

I'm calling `pickle.dumps(psi)` and store that in a postgres binary column, then I `psi = pickle.loads(psi_bytes)` and then when I pass psi into async_finish_signing I get the error you experienced:

```python
for index, (field_name, _, params) in enumerate(self._fields):
ValueError: not enough values to unpack (expected 3, got 2)
```
